File: cpp/ompl_planning.cpp
#include "ompl_planning.h"

#include "constants.h"
#include "ompl_validity_checker.h"
#include "ompl_helpers.h"

#include <drake/multibody/plant/multibody_plant.h>
#include <drake/planning/robot_diagram.h>

#include <ompl/base/spaces/RealVectorBounds.h>
#include <ompl/base/spaces/RealVectorStateSpace.h>
#include <ompl/base/SpaceInformation.h>
#include <ompl/base/ScopedState.h>
#include <ompl/base/ProblemDefinition.h>
#include <ompl/base/objectives/PathLengthOptimizationObjective.h>
#include <ompl/geometric/planners/rrt/RRTConnect.h>
#include <ompl/base/Planner.h>
#include <ompl/base/PlannerStatus.h>

#include <Eigen/Dense>

#include <optional>
#include <memory>

using drake::multibody::MultibodyPlant;
using drake::planning::RobotDiagram;

namespace ob = ompl::base;
namespace og = ompl::geometric;

namespace motion_planning {
namespace ompl {

std::optional<Eigen::MatrixXd> GenerateWaypoints(
    const MultibodyPlant<double>& plant,
    std::shared_ptr<RobotDiagram<double>> diagram,
    const Eigen::VectorXd q_goal
) {
    // create context
    auto context { diagram->CreateDefaultContext() };
    const auto& fixed_plant_context { plant.GetMyContextFromRoot(*context) };
    auto q_start { plant.GetPositions(fixed_plant_context) };

    // create space
    Eigen::VectorXd lower_bounds { plant.GetPositionLowerLimits() };
    Eigen::VectorXd upper_bounds { plant.GetPositionUpperLimits() };
    ob::RealVectorBounds bounds { constants::SO101_NUM_Q };
    for (int i { 0 }; i < constants::SO101_NUM_Q; ++i) {
        bounds.setLow(i, lower_bounds(i));
        bounds.setHigh(i, upper_bounds(i));
    }
    auto space { std::make_shared<ob::RealVectorStateSpace>(constants::SO101_NUM_Q) };
    space->setBounds(bounds);

    // create space information
    auto si { std::make_shared<ob::SpaceInformation>(space) };
    si->setStateValidityChecker(
        std::make_shared<DrakeSO101ValidityChecker>(si, diagram)
    );
    si->setup();

    // create problem definition
    ob::ScopedState start { space };
    ob::ScopedState goal { space };
    for (int i { 0 }; i < constants::SO101_NUM_Q; ++i) {
        start->as<ob::RealVectorStateSpace::StateType>()->values[i] = q_start(i);
        goal->as<ob::RealVectorStateSpace::StateType>()->values[i] = q_goal(i);
    }
    auto pdef { std::make_shared<ob::ProblemDefinition>(si) };
    pdef->setStartAndGoalStates(start, goal);
    pdef->setOptimizationObjective(
        std::make_shared<ob::PathLengthOptimizationObjective>(si)
    );

    // create planner
    ob::PlannerPtr planner { std::make_shared<og::RRTConnect>(si, true) };
    planner->setProblemDefinition(pdef);
    planner->setup();

    // init constants
    constexpr int max_iters = { 10 };
    constexpr int min_waypoints { 5 };
    constexpr int max_waypoints { 8 };
    constexpr double low_clearance { 1e-2 };
    
    // generate waypoints
    Eigen::MatrixXd waypoints_mat {};
    int num_waypoints {};
    bool waypoints_found { false };
    for (int iters { 0 }; iters < max_iters; ++iters) {
        // clear previous data
        planner->clear();
        pdef->clearSolutionPaths();
        
        // solve planning problem
        ob::PlannerStatus solved { planner->solve(1.0) };
        if (!solved) {
            continue;
        }
        auto* path { pdef->getSolutionPath()->as<og::PathGeometric>() };
        
        // prune path
        path->subdivide();
        std::vector<Eigen::VectorXd> waypoints {
            helpers::state_to_vector(path->getState(0))
        };
        num_waypoints = 1;
        for (int i { 1 }; i < path->getStateCount() - 1; ++i) {
            auto s { path->getState(i) };
            bool valid { si->getStateValidityChecker()->isValid(s) };
            double clearance { si->getStateValidityChecker()->clearance(s) };
            if (valid && (clearance < low_clearance)) {
                waypoints.push_back(helpers::state_to_vector(s));
                ++num_waypoints;
            }
        }
        waypoints.push_back(
            helpers::state_to_vector(path->getState(path->getStateCount() - 1))
        );
        ++num_waypoints;
        if ((num_waypoints >= min_waypoints) && (num_waypoints <= max_waypoints)) {
            waypoints_mat = Eigen::MatrixXd { constants::SO101_NUM_Q, num_waypoints };
            for (int i { 0 }; i < num_waypoints; ++i) {
                waypoints_mat.col(i) = waypoints[i];
            }
            waypoints_found = true;
            break;
        }
    }

    if (waypoints_found)
        return waypoints_mat;
    else
        return std::nullopt;
}

} // namespace ompl
} // namespace motion_planning

File: examples/python/jupyter/ompl_planning.py
import numpy as np
from ompl import base as ob
from ompl import geometric as og
from pydrake.all import (
    RobotDiagram,
    ModelInstanceIndex,
    CollisionCheckerParams,
    SceneGraphCollisionChecker,
    MultibodyPlant,
    BsplineBasis,
    BsplineTrajectory,
    KinematicTrajectoryOptimization,
    MinimumDistanceLowerBoundConstraint,
    Solve,
)


class DrakeStateValidityChecker(ob.StateValidityChecker):
    def __init__(
        self, 
        si: ob.SpaceInformation,
        diagram: RobotDiagram,
        robot_model: ModelInstanceIndex,
        num_q: int,
        q_start: np.ndarray
    ):
        super().__init__(si)

        self.collision_checker_params = CollisionCheckerParams()
        self.collision_checker_params.model = diagram
        self.collision_checker_params.robot_model_instances = [robot_model]
        self.collision_checker_params.edge_step_size = 0.01
        self.collision_checker = SceneGraphCollisionChecker(self.collision_checker_params)

        self.context = self.collision_checker.MakeStandaloneModelContext()

        self.num_q = num_q
        self.influence_distance = 1e0
        self.q_start = q_start

    def _stateToVector(self, state: ob.State) -> np.ndarray[np.float64]:
        return np.array([state[i] for i in range(self.num_q)], dtype=np.float64)

    def isValid(self, state):
        q = self._stateToVector(state)
        if np.allclose(q, self.q_start):
            return True
        return self.collision_checker.CheckContextConfigCollisionFree(
            self.context, q
        )


class SO101SamplingPlanner:

    @classmethod
    def generate_path(
        cls,
        diagram: RobotDiagram,
        q_start: np.ndarray,
        q_goal: np.ndarray
    ):
        plant = diagram.plant()
        so101 = plant.GetModelInstanceByName("so101_new_calib")

        num_q = plant.num_positions()
        lower_bounds = plant.GetPositionLowerLimits()
        upper_bounds = plant.GetPositionUpperLimits()

        bounds = ob.RealVectorBounds(num_q)
        for i in range(num_q):
            bounds.setLow(i, lower_bounds[i])
            bounds.setHigh(i, upper_bounds[i])
        bounds.setLow(5, np.pi/4)
        bounds.setHigh(5, np.pi/4)

        space = ob.RealVectorStateSpace(num_q)
        space.setBounds(bounds)

        si = ob.SpaceInformation(space)
        validityChecker = DrakeStateValidityChecker(
            si, diagram, so101, num_q, q_start
        )
        si.setStateValidityChecker(validityChecker)
        si.setup()

        start = space.allocState()
        goal = space.allocState()
        for i in range(num_q):
            start[i] = q_start[i]
            goal[i] = q_goal[i]

        pdef = ob.ProblemDefinition(si)
        pdef.setStartAndGoalStates(start, goal)
    
        planner = og.RRTConnect(si, addIntermediateStates=True)
        planner.setProblemDefinition(pdef)
        planner.setup()

        solved = planner.solve(1.0)
        if solved:
            path = pdef.getSolutionPath()
            num_states = path.getStateCount()
            dim = space.getDimension()

            path_array = np.zeros((num_states, dim))
            for i in range(num_states):
                state = path.getState(i)
                for j in range(dim):
                    path_array[i, j] = state[j]

            waypoints = path_array.T
            return waypoints
        else:
            return None

    @classmethod
    def generate_trajectory(
        cls, 
        plant: MultibodyPlant,
        diagram: RobotDiagram,
        waypoints: np.ndarray,
        avoid_collisions = True
    ) -> BsplineTrajectory | None:
        num_q = plant.num_positions()
        basis = BsplineBasis(4, waypoints.shape[1])
        init_traj = BsplineTrajectory(basis, waypoints)
        trajopt = KinematicTrajectoryOptimization(num_q, waypoints.shape[1], 4)
        trajopt.SetInitialGuess(init_traj)
    
        trajopt.AddDurationCost(1.0)
        trajopt.AddPathLengthCost(1.0)
        trajopt.AddPositionBounds(
            plant.GetPositionLowerLimits(), 
            plant.GetPositionUpperLimits()
        )
        trajopt.AddVelocityBounds(
            0.3 * plant.GetVelocityLowerLimits(),
            0.3 * plant.GetVelocityUpperLimits()
        )
        trajopt.AddAccelerationBounds(
            -2.0 * np.ones(num_q),
                2.0 * np.ones(num_q)
        )
        trajopt.AddJerkBounds(
            -1.0 * np.ones(num_q),
                1.0 * np.ones(num_q)
        )
        trajopt.AddDurationConstraint(0.5, 5.0)
    
        q_start = waypoints[:, 0]
        q_goal = waypoints[:, -1]
        trajopt.AddPathPositionConstraint(lb=q_start, ub=q_start, s=0)
        trajopt.AddPathPositionConstraint(lb=q_goal, ub=q_goal, s=1)
    
        trajopt.AddPathVelocityConstraint(np.zeros((num_q, 1)), np.zeros((num_q, 1)), 0)
        trajopt.AddPathVelocityConstraint(np.zeros((num_q, 1)), np.zeros((num_q, 1)), 1)

        so101 = plant.GetModelInstanceByName("so101_new_calib")
        if avoid_collisions:
            collision_checker_params = CollisionCheckerParams()
            collision_checker_params.model = diagram
            collision_checker_params.robot_model_instances = [so101]
            collision_checker_params.edge_step_size = 0.01
            collision_checker = SceneGraphCollisionChecker(collision_checker_params)
            # collision_checker.SetPaddingAllRobotEnvironmentPairs(1e-3)
            collision_constraint = MinimumDistanceLowerBoundConstraint(
                collision_checker,
                5e-3,
                collision_checker.MakeStandaloneModelContext(),
                None,
                5e-2,
            )
            evaluate_at_s = np.linspace(0, 1, 25)
            for s in evaluate_at_s:
                trajopt.AddPathPositionConstraint(collision_constraint, s)
    
        prog = trajopt.get_mutable_prog()
        result = Solve(prog)
        if result.is_success():
            return trajopt.ReconstructTrajectory(result)
        else:
            return None
